libs/common-services/src/main/java/uk/gov/di/ipv/core/library/persistence/InMemoryDataStore.java

```diff
@@ -0,0 +1,205 @@
+                    "Missing partition key from class " + klass.getName());
+        }
+
+        this.records =
+                (ConcurrentMap<String, T>)
+                        TABLES.computeIfAbsent(tableName, k -> new ConcurrentHashMap<>());
+        this.klass = klass;
+        this.partitionKeyMethod = partitionKey;
+        this.sortKeyMethod = sortKey;
+        this.hasSortKey = sortKey != null;
+    }
+
+    @Override
+    public void create(T item, ConfigurationVariable tableTtl) {
+        create(item);
+    }
+
+    @Override
+    public void create(T item) {
+        records.put(getKey(item), item);
+    }
+
+    @Override
+    public void createIfNotExists(T item) throws ItemAlreadyExistsException {
+        var key = getKey(item);
+        if (records.putIfAbsent(key, item) != null) {
+            throw new ItemAlreadyExistsException();
+        }
+    }
+
+    @Override
+    public T getItem(String partitionValue, String sortValue) {
+        return records.get(getKey(partitionValue, sortValue));
+    }
+
+    @Override
+    public T getItem(String partitionValue) {
+        return records.get(getKey(partitionValue, null));
+    }
+
+    @Override
+    public T getItem(String partitionValue, boolean warnOnNull) {
+        return getItem(partitionValue);
+    }
+
+    @Override
+    public T getItemByIndex(String indexName, String value) {
+        Method indexMethod = null;
+        for (var method : klass.getMethods()) {
+            if (method.isAnnotationPresent(DynamoDbSecondaryPartitionKey.class)) {
+                var annotation = method.getAnnotation(DynamoDbSecondaryPartitionKey.class);
+                if (Arrays.asList(annotation.indexNames()).contains(indexName)) {
+                    indexMethod = method;
+                }
+            }
+        }
+        if (indexMethod == null) {
+            throw new IllegalArgumentException("Missing index " + indexName);
+        }
+
+        for (var item : records.values()) {
+            try {
+                if (indexMethod.invoke(item).toString().equals(value)) {
+                    return item;
+                }
+            } catch (IllegalAccessException | InvocationTargetException e) {
+                throw new IllegalArgumentException("Could not access index " + indexName);
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public List<T> getItems(String partitionValue) {
+        return records.values().stream()
+                .filter(i -> getPartitionKey(i).equals(partitionValue))
+                .toList();
+    }
+
+    @SuppressWarnings(
+            "java:S3011") // We rely on field access to avoid needing to search for a getter
+    @Override
+    public List<T> getItemsWithBooleanAttribute(String partitionValue, String name, boolean value) {
+        try {
+            var field = klass.getDeclaredField(name);
+            field.setAccessible(true);
+            return records.values().stream()
```

Comment:
This doesn't seem to be using the partition value - we'd want to filter by that first I think. Otherwise you'll end up with a bunch of items you're not interested in?

Reply:
Good spot!